### PR TITLE
fix(ext-power): Initialize as soon as settings are available

### PR DESCRIPTION
**Pull request by Nicell**
This adjusts the initialization of the `ext_power` driver to be right after the flash systems have been initialized (All flash systems initialize by `POST_KERNEL` priority `80`). This way we know `ext_power` is properly initialized for the rest of the devices that need external power on start.

Currently, this fixes displays not turning on after deep sleep, but this should prevent similar issues in the future.

This is a step towards recovering devices when `ext_power` is in use. The next step is tying more devices to the PM policy so they can re-initialize without the entire system needing to be reset when `ext_power` is used while we keep context (idle, behaviors, etc).

I'm wondering if this init priority should be a Kconfig value.

Some boards' power switch takes some substantial time to actually switch to powered on, so the `init-delay-ms` property has been added, which holds the system for `n` milliseconds after enabling (or disabling) the power switch on initialization. This makes sure that when devices like the OLED are initialized, the power is ready to be used.

~~This is also not fully tested but works in theory. We need to reconfirm this change works and doesn't break the settings integration with `ext_power`~~ Re-init tested by @mcrosson, @tominabox1, and @karnadii. Settings confirmed to hold on start by @mcrosson.

### CODE REVIEW

**Comment by mcrosson** [comment]
Initial testing looks good.

I have *not* tried checking the settings but the oled *is* re-initializing when waking from deep sleep.

**Comment by tominabox1** [comment]
The additional delay fixed the issue I was having with my BLe Chiffre 👍🏼

**Comment by mcrosson** [comment]
bump? sounds like this is working all around for folks

**Comment by karnadii** [comment]
> The additional delay fixed the issue I was having with my BLe Chiffre 👍🏼

does this fix the wake up after deep sleep issue?

**Comment by tominabox1** [comment]
Yes!

**Comment by karnadii** [comment]
it is definitely fixes problem wake up after sleep. but when ext vcc is cut off the oled still not reinit. ext power toogle and rgb underglow still won't turn oled back on.

**Comment by Nicell** [comment]
> it is definitely fixes problem wake up after sleep (the oled reinit and no freeze after wake up). but when ext vcc is cut off the oled still not reinit. ext power toggle and rgb underglow toggle still won't turn oled back on.

This is known and is a separate issue than the one this fixes. Power management policy related items need to be added to drivers that make use of ext power, but for now this fixes ext power not being on early enough during full initialization.

**Comment by karnadii** [comment]
> > it is definitely fixes problem wake up after sleep (the oled reinit and no freeze after wake up). but when ext vcc is cut off the oled still not reinit. ext power toggle and rgb underglow toggle still won't turn oled back on.
> 
> This is known and is a separate issue than the one this fixes. Power management policy related items need to be added to drivers that make use of ext power, but for now this fixes ext power not being on early enough during full initialization.

ok, seems I misunderstand the PR. at least we have deep sleep now. I won't worry about my battery now. thanks for the PR.

**Comment by karnadii** [comment]
I have tried on my marvelous65-split using nrfmicro 1.3. but it did not work, the oled did not re-init and after a while it freeze. while my marvelous65 (nrfmicro 1.4, non-split) worked. Please someone with split keyboard confirm this.

**Comment by tominabox1** [comment]
in your .dts file you have to add the appropriate init delay

`ext-power {
		compatible = "zmk,ext-power-generic";
		label = "EXT_POWER";
		control-gpios = <&gpio0 7 GPIO_ACTIVE_LOW>;
		init-delay-ms = <250>;
	};`

**Comment by karnadii** [comment]
> in your .dts file you have to add the appropriate init delay
> 
> `ext-power { compatible = "zmk,ext-power-generic"; label = "EXT_POWER"; control-gpios = <&gpio0 7 GPIO_ACTIVE_LOW>; init-delay-ms = <250>; };`

what is this `control-gpios = <&gpio0 7 GPIO_ACTIVE_LOW>`? is it like the underglow will be ignored?

**Comment by tominabox1** [comment]
> > in your .dts file you have to add the appropriate init delay
> > `ext-power { compatible = "zmk,ext-power-generic"; label = "EXT_POWER"; control-gpios = <&gpio0 7 GPIO_ACTIVE_LOW>; init-delay-ms = <250>; };`
> 
> what is this `control-gpios = <&gpio0 7 GPIO_ACTIVE_LOW>`? is it like the underglow will be ignored? unused pin?

in the .dts file for your keyboard, if your board supports ext power toggle, there will be this section. With this PR, you have to add the init-delay-ms line to that block of code.

File: app/src/ext_power_generic.c [hunk 1]
```diff
@@ -210,7 +215,8 @@ static int ext_power_generic_pm_control(const struct device *dev, uint32_t ctrl_
 static const struct ext_power_generic_config config = {
     .label = DT_INST_GPIO_LABEL(0, control_gpios),
     .pin = DT_INST_GPIO_PIN(0, control_gpios),
-    .flags = DT_INST_GPIO_FLAGS(0, control_gpios)};
+    .flags = DT_INST_GPIO_FLAGS(0, control_gpios),
+    COND_CODE_0(DT_INST_NODE_HAS_PROP(0, init_delay_ms), (0), (DT_INST_PROP(0, init_delay_ms)))};
```
Comment (petejohanson) [inline app/src/ext_power_generic.c:219]:
```suggestion
    .init_delay_ms = DT_INST_PROP_OR(0, init_delay_ms, 0)};
```

Use `DT_INST_PROP_OR`. I know there's other code that uses this approach, but it was before I found that helper.

File: app/src/ext_power_generic.c [hunk 2]
```diff
@@ -223,13 +229,15 @@ static const struct ext_power_api api = {.enable = ext_power_generic_enable,
                                          .disable = ext_power_generic_disable,
                                          .get = ext_power_generic_get};
 
+#define ZMK_EXT_POWER_INIT_PRIORITY 81
```
Comment (petejohanson) [inline app/src/ext_power_generic.c:232]: This is fine for now, but something to consider moving to Kconfig as needed.

**Review by petejohanson — APPROVED** [review]
Thanks for digging into this!

**Pull request merged**